internal/util/password/scramsha1.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package password
+
+import (
+	"crypto/md5"
+	"crypto/rand"
+	"crypto/sha1"
+	"encoding/hex"
+
+	"golang.org/x/crypto/pbkdf2"
+
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+)
+
+// SCRAMSHA1Hash computes SCRAM-SHA-1 credentials and returns the document that should be stored.
+func SCRAMSHA1Hash(username, password string) (*types.Document, error) {
```

Comment:
Would we need verify SHA1 function or equivalent?

Currently Conv is created sasl start, then sasl continue checks conversation. Let's say some other user updates the user's password or delete the user. For such case, what is the expected outcome? I think having SHA1 verification similar to PLAIN mechanism is sensible. :hugs:

Reply:
This is a but confusion, but it works this way:

Whenever the [Step function](https://github.com/xdg-go/scram/blob/17629a50d5ce12875d83f9095809ae43b765c303/server_conv.go#L46-L58) is invoked, the callback function scramCredentialLookup is called by Step (see [ServerConversation.credentialCB](https://github.com/xdg-go/scram/blob/17629a50d5ce12875d83f9095809ae43b765c303/server_conv.go#L95)), except for the final message. This is defined at FerretDB's code on the saslStartSCRAM function.

So the user is authenticated at each step.

Reply:
My confusion is cleared :+1: